sdk/Sdk/FunctionMetadataGenerator.cs

```diff
@@ -378,14 +390,217 @@ private static ExpandoObject BuildBindingMetadataFromAttribute(CustomAttribute a
             bindingDict["Type"] = bindingType;
             bindingDict["Direction"] = GetBindingDirection(attribute);
 
+            // Is string parameter type
+            if (IsStringType(parameterType.FullName))
+            {
+                bindingDict["DataType"] = "String";
+            }
+            // Is binary parameter type
+            else if (IsBinaryType(parameterType.FullName))
+            {
+                bindingDict["DataType"] = "Binary";
+            }
+
             foreach (var property in attribute.GetAllDefinedProperties())
             {
                 bindingDict.Add(property.Key, property.Value);
             }
 
+            // TODO: do not rely on property alone
+            if (bindingDict["isBatched"] is not null
+                && bindingDict["isBatched"] is bool isBatchedValue)
+            {
+                // Batching set to true
+                if (isBatchedValue)
+                {
+                    bindingDict["Cardinality"] = "Many";
+                    // Throw if parameter type is not IEnumerable
+                    if (IsIterableCollection(parameterType, out DataType dataType))
+                    {
+                        if (dataType.Equals(DataType.String))
+                        {
+                            bindingDict["DataType"] = "String";
+                        }
+                        else if (dataType.Equals(DataType.Binary))
+                        {
+                            bindingDict["DataType"] = "Binary";
+                        }
+                    }
+                    else
+                    {
+                        throw new Exception("Function is configured to process events in batches but parameter type is not iterable. " +
+                            $"Change parameter { parameterName ?? "type" } to be an IEnumerable type or set 'IsBatched' to false on your {attribute.AttributeType.Name.Replace("Attribute", "")} attribute.");
+                    }
+                }
+                // Batching set to false
+                else
+                {
+                    bindingDict["Cardinality"] = "One";
+                }
+
+                bindingDict.Remove("isBatched");
+            }
+
             return binding;
         }
 
+        private static bool IsIterableCollection(TypeReference type, out DataType dataType)
+        {
+            // Array and not byte array 
```

Comment:
Unfortunately no :( when we "Resolve" the type reference to a type definition, it just resolves the non-array element type. I tried removing this too and the `byte[][]` test fails. Interestingly, `string[]` doesn't fail, but only because string is an ienumerable<char>